gdb-server/src/architecture.rs

```diff
@@ -0,0 +1,78 @@
+use probe_rs::{Core, CoreRegisterAddress};
+
+/// Extension trait for probe_rs::Core, which adds some GDB -> probe-rs internal translation functions
+///
+/// translates some GDB architecture dependant stuff
+/// to probe-rs internals.
+pub(crate) trait GdbArchitectureExt {
+    /// Translate GDB register number to internal register address
+    fn translate_gdb_register_number(
+        &self,
+        gdb_reg_number: u32,
+    ) -> Option<(CoreRegisterAddress, u32)>;
+
+    /// Number of general registers
+    fn num_general_registers(&self) -> usize;
+}
+
+impl<'probe> GdbArchitectureExt for Core<'probe> {
+    fn translate_gdb_register_number(
+        &self,
+        gdb_reg_number: u32,
+    ) -> Option<(CoreRegisterAddress, u32)> {
+        let (probe_rs_number, bytesize): (u16, _) = match self.architecture() {
+            probe_rs::Architecture::Arm => {
+                match gdb_reg_number {
+                    // Default ARM register (arm-m-profile.xml)
+                    // Register 0 to 15
+                    x @ 0..=15 => (x as u16, 4),
+                    // CPSR register has number 16 in probe-rs
+                    // See REGSEL bits, DCRSR register, ARM Reference Manual
+                    25 => (16, 4),
+                    // Floating Point registers (arm-m-profile-with-fpa.xml)
+                    // f0 -f7 start at offset 0x40
+                    // See REGSEL bits, DCRSR register, ARM Reference Manual
+                    reg @ 16..=23 => ((reg as u16 - 16 + 0x40), 12),
+                    // FPSCR has number 0x21 in probe-rs
+                    // See REGSEL bits, DCRSR register, ARM Reference Manual
+                    24 => (0x21, 4),
+                    // Other registers are currently not supported,
+                    // they are not listed in the xml files in GDB
+                    other => {
+                        log::warn!("Request for unsupported register with number {}", other);
+                        return None;
+                    }
+                }
+            }
+            probe_rs::Architecture::Riscv => match gdb_reg_number {
+                // general purpose registers 0 to 31
+                x @ 0..=31 => {
+                    let addr: CoreRegisterAddress = self
+                        .registers()
+                        .get_platform_register(x as usize)
+                        .expect("riscv register must exist")
+                        .into();
+                    (addr.0, 8)
```

Comment:
Do you know why these registers have to be 64-bit for GDB? Based on the [XML](https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=gdb/features/riscv/32bit-cpu.xml;h=b75f0b11606be8a4ae00f46bd80c11caa0e36765;hb=HEAD) file I would have expected GDB to support 32-bit registers.

I wonder if we have to tell GDB the architecture of the target, as done for ARM in the `host_info` function:

https://github.com/probe-rs/probe-rs/blob/23cd914255d25b628deaf53a5983901158f806f4/gdb-server/src/handlers.rs#L147-L152

Reply:
ohhh i didnt realize this was hardcoded! i tried changing the architecture from within GDB for thumbv7 targets :P

Reply:
I'm actually not sure if this is only used by LLVM, there are also XML-based target descriptions supported by GDB: https://www.sourceware.org/gdb/current/onlinedocs/gdb/Target-Descriptions.html#Target-Descriptions

Reply:
I am 99% sure GDB uses this :)

Reply:
At least my GDB (version 9.2) shows the following architecture, when debugging a binary for a Cortex-M4:

```
(gdb) show architecture 
The target architecture is set automatically (currently armv7e-m)
```
So it either ignores the `host_info` output, or doesn't use it. When no binary is specified, it show `i386` as architecture.

Reply:
I tried with several builds of gdb, for riscv, for riscv32, all of them expect 64bit based on testing.
Not sure why, maybe they decided to handle riscv generically. But what about the 128bit ISA? Confuses me as well.

Reply:
I will update the host_info() function

Reply:
I've created an issue (#495)  for target description support, that might be useful to add at some point. For now using 64 bit registers seems fine to me.

Reply:
> At least my GDB (version 9.2) shows the following architecture, when debugging a binary for a Cortex-M4:
> 
> ```
> (gdb) show architecture 
> The target architecture is set automatically (currently armv7e-m)
> ```
> 
> So it either ignores the `host_info` output, or doesn't use it. When no binary is specified, it show `i386` as architecture.

Hmm for us this was always `armv6-none-eabi` when debugging an STM32L4 (M4, armv7) and we also wernt able to force GDB to change it.